core/src/odbc_uri.rs

```diff
@@ -148,36 +152,70 @@ impl<'a> ODBCUri<'a> {
     fn remove_mandatory_attribute(&mut self, names: &[&str]) -> Result<&'a str> {
         self.remove(names).ok_or_else(|| {
             if names.len() == 1 {
-                ODBCError::InvalidUriFormat(format!(
+                Error::InvalidUriFormat(format!(
                     "{} is required for a valid Mongo ODBC Uri",
                     names[0]
                 ))
             } else {
-                ODBCError::InvalidUriFormat(format!(
+                Error::InvalidUriFormat(format!(
                     "One of {:?} is required for a valid Mongo ODBC Uri",
                     names
                 ))
             }
         })
     }
 
-    // remove_to_mongo_uri converts this ODBCUri to a mongo_uri String. It will
+    // try_into_client_options converts this ODBCUri to a mongo_uri String. It will
     // remove all the attributes necessary to make a mongo_uri. This is destructive!
-    pub fn remove_to_mongo_uri(&mut self) -> Result<String> {
+    pub fn try_into_client_options(&mut self) -> Result<ClientOptions> {
+        let uri = self.remove(URI);
+        if let Some(uri) = uri {
+            return self.handle_uri(uri);
+        }
+        self.handle_no_uri()
+    }
+
+    fn handle_uri(&mut self, uri: &str) -> Result<ClientOptions> {
+        let mut client_options = ClientOptions::parse(uri)?;
```

Comment:
The authSource actually won't be lost in this code. If there is a URI, the only thing it does now is _update_ the `ClientOptions` with credentials. None of the original parsed settings are lost.

I should add a test for that, however!.

Reply:
I don't think so. Yes,please add a test, because I don't understand how it can work.
Here is the code from my own side test:
```
fn check_cred(client_options: &ClientOptions) {
    if client_options.credential.is_some() {
        let creds = (client_options.clone()).credential.unwrap();
        if creds.source.is_some() {
            println!("AuthSource = {}", creds.source.unwrap());
        } else {
            println!("No authSource");
        }
    } else {
        println!("No credential")
    }
}
#[test]
fn test_client_options() {
    let uris = vec![
        "mongodb://localhost/?authSource=authDB",
        "mongodb://localhost/",
        "mongodb://uid@pwdlocalhost/?authSource=authDB",
        "mongodb://uid@pwdlocalhost/",
    ];
    println!("---------------------------------");
    for uri in uris {
        println!("-- URI : {} --", &uri);
        let mut client_options = ClientOptions::parse(uri).unwrap();

        print!("Initial parse ==> ");
        check_cred(&client_options);

        if client_options.credential.is_none() {
            print!("Adding UID and PWD retrieved from somewhere else ==> ");
            // Add credential info
            client_options.credential = Some(
                Credential::builder()
                    .username("uid".to_string())
                    .password("pwd".to_string())
                    .build(),
            );

            check_cred(&client_options);
        }
        println!("---------------------------------");
    }
}
```

and here is the output :
```
---------------------------------
-- URI : mongodb://localhost/?authSource=authDB --
Initial parse ==> No credential
Adding UID and PWD retrieved from somewhere else ==> No authSource
---------------------------------
-- URI : mongodb://localhost/ --
Initial parse ==> No credential
Adding UID and PWD retrieved from somewhere else ==> No authSource
---------------------------------
-- URI : mongodb://uid@pwdlocalhost/?authSource=authDB --
Initial parse ==> AuthSource = authDB
---------------------------------
-- URI : mongodb://uid@pwdlocalhost/ --
Initial parse ==> AuthSource = admin
---------------------------------
```

Because the code replaces the credential sub-struct with the one create in the function which only assigns username and password, it looks like there will be no authsource set if the original URI does not contains a UID + PWD.

Reply:
It doesn't replace it, it updates it. The problem is the URI parser doesn't create a `Credential` when there is an `authSrc` but no `UID` or `PWD`